go/oasis-test-runner/scenario/e2e/runtime/trust_root_change.go

```diff
@@ -0,0 +1,444 @@
+//
+// It consists of 3 steps:
+//   - Build a simple key/value runtime with an embedded trust root, register
+//     it to the network together with key manager runtime and test that
+//     everything works.
+//   - Stop the network and reset the consensus state. This will cause a chain
+//     context change when the network will be started again. When doing state
+//     wipe be careful not to delete compute workers local storages as they
+//     contain sealed trusted roots. We also remove one validator from the set
+//     so that we can simulate what happens when the new validator set has only
+//     2/3 of the voting power.
+//   - Start the network. If everything works as expected, compute workers
+//     should never be ready as the verifier cannot transfer trust to light
+//     blocks on a new chain.
+func (sc *trustRootChangeImpl) unhappyRun(childEnv *env.Env) (err error) {
+	ctx := context.Background()
+
+	// Step 1: Build a simple key/value runtime and start the network.
+	rebuild, err := sc.buildKeyValueRuntime(ctx, childEnv)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err2 := rebuild(); err2 != nil {
+			err = fmt.Errorf("%w (original error: %s)", err2, err)
+		}
+	}()
+
+	chainContext, err := sc.chainContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	lastHeight, err := sc.latestHeight(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Step 2: Stop and reset the network.
+	if err = sc.resetNetwork(childEnv); err != nil {
+		return err
+	}
+
+	// Step 3: Start the network and verify compute workers.
+	if err = sc.Net.Start(); err != nil {
+		return err
+	}
+	if err = sc.Net.Controller().WaitNodesRegistered(ctx, len(sc.Net.Validators())); err != nil {
+		return err
+	}
+	if _, err = sc.waitBlocks(ctx, 3); err != nil {
+		return err
+	}
+
+	// Assert that chain context has changed. Test is meaningless if this
+	// doesn't happen.
+	newChainContext, err := sc.chainContext(ctx)
+	if err != nil {
+		return err
+	}
+	if newChainContext == chainContext {
+		return fmt.Errorf("chain context hasn't changed")
+	}
+
+	// Running network to approximately the same height as in the first run
+	// should be enough to collect both verifier's errors.
+	sc.Logger.Info("running network to approximately the same hight as before")
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	go func() {
+		if _, err := sc.waitBlocks(ctx, int(lastHeight)); err == nil {
+			cancel()
+		}
+	}()
+
+	// Compute workers should have problems starting the key/value runtime
+	// as the verifier will never initialize.
+	sc.Logger.Info("starting compute workers")
+	for _, n := range sc.Net.ComputeWorkers() {
+		if err := n.Start(); err != nil {
+			return fmt.Errorf("failed to start node: %w", err)
+		}
+	}
+	for _, n := range sc.Net.ComputeWorkers() {
+		if err := n.WaitReady(ctx); err != nil {
+			if status, ok := status.FromError(err); ok {
+				switch status.Code() {
+				case codes.Canceled, codes.DeadlineExceeded:
```

Comment:
How long will this wait in practice? The full 5 minutes? This could make the test overly long so it would be great if this could be reduced.